File: clibs/low-utils/trie/writer.cs
using System;
using System.Collections.Generic;

public class BytesWriter {

  List<byte[]> bytes = new List<byte[]>();

  public int len;

  public void Add(int num, byte size /*0,1,2,3*/) {
    if (num == 0) return;
    Add(
      size == 1 ? new byte[] { (byte)num } : (
      size == 2 ? new byte[] { (byte)(num & 0xFF), (byte)((num >> 8) & 0xFF) } :
      ((UInt24)num).buf));
  }
  public void Add(byte[] data) {
    if (data == null) return;
    len += data.Length;
    bytes.Add(data);
  }

  public void Add(BytesWriter data) {
    if (data == null) return;
    len += data.len;
    bytes.AddRange(data.bytes);
  }

  public byte[] toBytes() {
    var res = new byte[len];
    int pos = 0;
    foreach (var bs in bytes) {
      Buffer.BlockCopy(bs, 0, res, pos, bs.Length);
      pos += bs.Length;
    }
    if (pos != len) throw new Exception("pos != len");
    return res;
  }

  public static byte getNumberSizeMask(int num) { // returns 0,1,2 or 3
    if (num > 0xffffff || num < 0) throw new ArgumentOutOfRangeException();
    return (byte)(num == 0 ? 0 : (num <= 0xff ? 1 : (num <= 0xffff ? 2 : 3)));
  }

}

File: dlibs/low-utils/lib/src/trie/writer.dart
import 'dart:typed_data';
import 'package:convert/convert.dart' as convert;

class BytesWriter {
  static int getNumberSizeMask(int number) {
    // returns 0,1,2 or 3
    if (number == null) return 0;
    if (number > 0xffffff || number < 0) throw ArgumentError();
    return number == 0 ? 0 : (number <= 0xff ? 1 : (number <= 0xffff ? 2 : 3));
  }

  int len = 0;
  List<Uint8List> _bytes = List<Uint8List>();

  void writeNumber(int number, int size /*0,1,2,3*/) {
    if (size == 0) return;
    writeBytes(Uint8List.fromList(size == 1
        ? [number]
        : (size == 2
            ? [number & 0xFF, (number >> 8) & 0xFF]
            : [number & 0xFF, (number >> 8) & 0xFF, (number >> 16) & 0xFF])));
  }

  // void clear() {
  //   _bytes = List<Uint8List>();
  //   len = 0;
  // }

  void writeByte(int byte) {
  }


  void writeBytes(Uint8List data) {
    if (data == null) return;
    len += data.length;
    _bytes.add(data);
  }

  void writeList(List<int> data) {
    writeBytes(Uint8List.fromList(data));
  }

  void writeWriter(BytesWriter data) {
    if (data == null) return;
    len += data.len;
    _bytes.addAll(data._bytes);
  }

  String hexDump() {
    return convert.hex.encode(toBytes());
  }

  Uint8List toBytes() {
    final res = Uint8List(len);
    int pos = 0;
    for (final bs in _bytes) {
      res.setAll(pos, bs);
      pos += bs.length;
    }
    if (pos != len) throw Exception("pos != len");
    return res;
  }
}
